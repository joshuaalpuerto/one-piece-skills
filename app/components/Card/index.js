import React from 'react'
import PropTypes from 'prop-types'
import { IdWrapper, ContentWrapper, Remove, Wrapper } from './styled'

/**
 * Make renderDelete available for them to customize or not to provide just for display
 * @param {*} param0
 */
function Card({ id, color, heading, subheading, renderDelete }) {
  return (
    <Wrapper>
      <IdWrapper bgColor={color}>{id}</IdWrapper>
      <ContentWrapper>
        {heading}
        {subheading && <span> {subheading} </span>}
      </ContentWrapper>
      <Remove>{renderDelete && renderDelete()}</Remove>
    </Wrapper>
  )
}

Card.defaultProps = {
  color: '#80878B',
  renderDelete: null
}

Card.propTypes = {
  id: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,
  heading: PropTypes.string.isRequired,
  renderDelete: PropTypes.func,
  subheading: PropTypes.string,
  color: PropTypes.string
}

export default Card
