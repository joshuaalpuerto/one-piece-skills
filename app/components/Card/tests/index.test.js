import React from 'react'
import { mount } from 'enzyme'

import 'jest-styled-components'

import Card from '../index'
import { IdWrapper, ContentWrapper, Remove } from '../styled'

describe('<Card />', () => {
  describe('<IdWrapper />', () => {
    const props = {
      id: 1,
      heading: 'React',
      subheading: '1 year'
    }

    it('should be background default #000', () => {
      const renderedComponent = mount(<Card {...props} />)
      const instance = renderedComponent.find(IdWrapper)

      expect(instance).toMatchSnapshot()
      expect(instance).toHaveStyleRule('background-color', '#000')
    })

    it('should have different background color if passes as prop', () => {
      const subProps = {
        ...props,
        color: 'lightgreen'
      }
      const renderedComponent = mount(<Card {...subProps} />)
      const instance = renderedComponent.find(IdWrapper)

      expect(instance).toMatchSnapshot()
      expect(instance).toHaveStyleRule('background-color', 'lightgreen')
    })

    it('should have id existing', () => {
      const renderedComponent = mount(<Card {...props} />)
      const instance = renderedComponent.find(IdWrapper)
      expect(instance.text()).toEqual('1')
    })
  })

  describe('<ContentWrapper />', () => {
    const props = {
      id: 1,
      heading: 'React',
      subheading: '1 year'
    }

    it('should exist', () => {
      const renderedComponent = mount(<Card {...props} />)
      const instance = renderedComponent.find(ContentWrapper)
      expect(instance.length).toEqual(1)
    })

    it('should not render span if subheading is null', () => {
      const subProps = {
        ...props,
        subheading: null
      }
      const renderedComponent = mount(<Card {...subProps} />)
      const wrapper = renderedComponent.find(ContentWrapper)
      const span = wrapper.find('span')
      expect(span.exists()).toEqual(false)
    })
  })

  describe('<Remove />', () => {
    const props = {
      id: 1,
      heading: 'React',
      subheading: '1 year'
    }

    it('should exist', () => {
      const renderedComponent = mount(<Card {...props} />)
      const instance = renderedComponent.find(Remove)
      expect(instance.length).toEqual(1)
    })

    it('should render nothing when renderDelete is null', () => {
      const renderedComponent = mount(<Card {...props} />)
      const instance = renderedComponent.find(Remove)
      expect(instance.text()).toEqual('')
    })

    it('should render  renderDelete if Exist', () => {
      const DeleteComponent = () => <div> Remove </div>
      const subProps = {
        ...props,
        renderDelete: DeleteComponent
      }
      const renderedComponent = mount(<Card {...subProps} />)
      const instance = renderedComponent.find(Remove)

      expect(instance.equals(<Remove>{DeleteComponent()}</Remove>)).toEqual(
        true
      )
    })
  })
})
