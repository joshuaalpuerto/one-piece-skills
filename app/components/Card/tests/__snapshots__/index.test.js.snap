// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> <IdWrapper /> should be background default #24333C 1`] = `
.c1 {
  background-color: #24333C;
  color: #fff;
  padding: 5px;
}

.c2:hover .c0 {
  background-color: #1b998e;
}

<styled__IdWrapper
  bgColor="#24333C"
>
  <StyledComponent
    bgColor="#24333C"
    forwardedClass={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": undefined,
        "componentStyle": ComponentStyle {
          "componentId": "c0",
          "isStatic": false,
          "lastClassName": "c1",
          "rules": Array [
            "background-color:",
            [Function],
            ";color:#fff;padding:5px;",
          ],
        },
        "displayName": "styled__IdWrapper",
        "render": [Function],
        "styledComponentId": "c0",
        "target": "div",
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      }
    }
    forwardedRef={null}
  >
    <div
      className="c0 c1"
    >
      1
    </div>
  </StyledComponent>
</styled__IdWrapper>
`;

exports[`<Card /> <IdWrapper /> should have different background color if passes as prop 1`] = `
.c1 {
  background-color: lightgreen;
  color: #fff;
  padding: 5px;
}

.c2:hover .c0 {
  background-color: #1b998e;
}

<styled__IdWrapper
  bgColor="lightgreen"
>
  <StyledComponent
    bgColor="lightgreen"
    forwardedClass={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": undefined,
        "componentStyle": ComponentStyle {
          "componentId": "c0",
          "isStatic": false,
          "lastClassName": "c1",
          "rules": Array [
            "background-color:",
            [Function],
            ";color:#fff;padding:5px;",
          ],
        },
        "displayName": "styled__IdWrapper",
        "render": [Function],
        "styledComponentId": "c0",
        "target": "div",
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      }
    }
    forwardedRef={null}
  >
    <div
      className="c0 c1"
    >
      1
    </div>
  </StyledComponent>
</styled__IdWrapper>
`;
